provd/internal/services/user/user.go

```diff
@@ -0,0 +1,293 @@
+	}
+	realName := user.GetRealName()
+	if realName == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "recieved an empty realName")
+	}
+	isAdmin := req.GetIsAdmin()
+	var accountType int32
+	if isAdmin {
+		accountType = 1
+	} else {
+		accountType = 0
+	}
+	password := user.GetPassword() // allow empty password?
+	autologin := user.GetAutoLogin()
+	hostname := user.GetHostname()
+	if hostname == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "recieved an empty hostname")
+	}
+	// Create the user
+	var userObjectPath dbus.ObjectPath
+	call := s.accounts.Call("org.freedesktop.Accounts.CreateUser", 0, username, realName, accountType)
+
+	err := call.Store(&userObjectPath)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create user: %s", err)
+	}
+
+	hashed, err := hashPassword(password)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to generate hashed password: %s", err)
+	}
+
+	// Set the password for the user
+	userObject := s.userFactory.GetUserObject(userObjectPath)
+	fmt.Printf(hashed)
+	err = userObject.Call("org.freedesktop.Accounts.User.SetPassword", 0, hashed, "").Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set password: %s", err)
+	}
+
+	// Set autologin for the user
+	err = userObject.Call("org.freedesktop.Accounts.User.SetAutomaticLogin", 0, autologin).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set autologin: %s", err)
+	}
+
+	// Set the hostname
+	err = s.hostname.Call("org.freedesktop.hostname1.SetStaticHostname", 0, hostname, false).Err
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to set hostname: %s", err)
+	}
+
+	return &proto.Empty{}, nil
+}
+
+func (s *Service) ValidateUsername(ctx context.Context, req *proto.ValidateUsernameRequest) (*proto.ValidateUsernameResponse, error) {
+	// Validate request
+	if req == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "received a nil request")
+	}
+	username := req.GetUsername()
+	if username == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "username cannot be empty")
+	}
+
+	err := s.accounts.Call("org.freedesktop.Accounts.FindUserByName", 0, username).Err
+	if err != nil {
+
+		// Check if the error is due to user not being found or a D-Bus error
+		if dbusError, ok := err.(dbus.Error); ok {
+			if dbusError.Name == "org.freedesktop.Accounts.Error.Failed" {
+				// User not found
+				return &proto.ValidateUsernameResponse{Valid: true}, nil
+			}
+			// Handle other dbus errors
+			return nil, status.Errorf(codes.Internal, "dbus error: %v", dbusError)
+		}
+		// Unknown error
+		return nil, status.Errorf(codes.Internal, "unknown error: %v", err)
+	}
+
+	// User found
+	return &proto.ValidateUsernameResponse{Valid: false}, nil
+
+}
+
+func (s *Service) GetUser(ctx context.Context, req *proto.GetUserRequest) (*proto.GetUserResponse, error) {
```

Comment:
As far as I understand we don't really need this method (and could even remove it from the protos).

Subiquity's `GetIdentity()` doesn't take any arguments and just returns an identity if it has already been configured before.

Reply:
Cool, I'll clean it up when I do my final code pass :slightly_smiling_face: